heksher/db_logic/rule.py

```diff
@@ -0,0 +1,222 @@
+from operator import itemgetter
+from typing import Dict, Optional, Any, List, Tuple, NamedTuple, Sequence
+
+import orjson
+from sqlalchemy import select, join
+
+from heksher.db_logic.logic_base import DBLogicBase
+from heksher.db_logic.metadata import rules, conditions, settings, context_features
+
+
+# These two classes should be used in the context of db logic api only
+
+class QueryResult(NamedTuple):
+    """
+    The result of a rule query
+    """
+    applicable_settings: List[str]
+    rules: List[RuleSpec]
+
+
+class RuleSpec(NamedTuple):
+    """
+    A single rule that results from a rule query or lookup
+    """
+    setting: str
+    value: Any
+    feature_values: Sequence[Tuple[str, str]]
+    metadata: Optional[Dict[str, Any]]
+
+
+class RuleMixin(DBLogicBase):
+    async def get_rule(self, id_: int) -> Optional[RuleSpec]:
+        """
+        Args:
+            id_: the id of a specific rule
+
+        Returns:
+            A RuleSpec describing the rule with the id, or None if no such rule exists.
+        """
+        basic_results, feature_values = await gather(
+            self.db.fetch_one(
+                select([rules.c.setting, rules.c.value, rules.c.metadata]).where(rules.c.id == id_)
+            ),
+            self.db.fetch_all(
+                select([conditions.c.context_feature, conditions.c.feature_value])
+                .select_from(join(conditions, context_features,
+                                  conditions.c.context_feature == context_features.c.name))
+                .where(conditions.c.rule == id_)
+                .order_by(context_features.c.index)
+            )
+        )
+        # we query both the rule and its feature values at the same time. Meaning that if the rule does not exist,
+        # we make 1 too many calls. However, we expect to make so few get_rule calls to non-existent rules that this
+        # is negligible
+        if not basic_results:
+            return None
+        metadata_ = orjson.loads(basic_results['metadata'])
+        value_ = orjson.loads(basic_results['value'])
+        return RuleSpec(
+            basic_results['setting'],
+            value_,
+            [(f['context_feature'], f['feature_value']) for f in feature_values],
+            metadata_
+        )
+
+    async def get_rule_id(self, setting: str, match_conditions: Dict[str, str]) -> Optional[int]:
+        """
+        Lookup a rule by its settings and conditions, and retrieve its id
+
+        Args:
+            setting: The name of the setting the rule pertains to
+            match_conditions: The exact-match conditions of the rule
+
+        Returns:
+            The id of the rule, or None if it does not exist
+
+        """
+        condition_count = len(match_conditions)
+        # we convert the conditions to a format that they can be queried with an "IN" condition
+        condition_tuples = ','.join(f"('{k}','{v}')" for (k, v) in match_conditions.items())
+        query_template = f'''
```

Comment:
you're sending the comments, which is redundant. The documentation is great, but you need to find a better suiting structure. Maybe closing the quotes each time and using a Python comment?